jte-kotlin/src/test/java/gg/jte/compiler/kotlin/TestKotlinParamInfo.java

```diff
@@ -0,0 +1,103 @@
+package gg.jte.compiler.kotlin;
+
+import gg.jte.compiler.ParamInfo;
+import gg.jte.compiler.TemplateParserVisitor;
+import gg.jte.compiler.TemplateParserVisitorAdapter;
+import org.junit.jupiter.api.Assertions;
+import org.junit.jupiter.api.Test;
+
+public class TestKotlinParamInfo {
```

Comment:
There is some repetition in the tests here, but I prefer it that way since they are easier to read (IMO). Let me know if you want me to refactor it somehow to better match the project's style.

Reply:
Got it. So, this PR introduces a test for _regular_ (non-Content) params with default values:

https://github.com/casid/jte/pull/285/files#diff-b16684272d3912ace7de14f3b2499de85c89003e7ee137ebd4af72bf89e41fc2R721

The two last points were addressed by 3e708872665f6ece25634e8e368c3d8ae18e6ff8.

Reply:
Hum, got a few unexpected test failures:

https://github.com/casid/jte/actions/runs/6485276269/job/17611002039?pr=285#step:5:312

@casid, I can investigate those later, but if you know what those are about, that can speed up my investigation.

Reply:
So, considering this test:

```java
    @Test
    void tagWithDefaultContentParam() {
        givenTag("tagWithDefaultContentParam", "@param one:Int = 1\n" +
                          "@param content:gg.jte.Content = @`Some Content`\n" +
                          "${one}, ${content}");
        givenTemplate("@template.tag.tagWithDefaultContentParam()");

        thenOutputIs("1, Some Content");
    }
```

The code generated for the tag when running `gg.jte.kotlin.TemplateEngine_BinaryTest` was:

```kotlin
@file:Suppress("ktlint")
package gg.jte.generated.ondemand.tag
@Suppress("UNCHECKED_CAST", "UNUSED_PARAMETER")
class JtetagWithDefaultContentParamGenerated {
    companion object {
    	@JvmField val JTE_NAME = "tag/tagWithDefaultContentParam.kte"
    	@JvmField val JTE_LINE_INFO = intArrayOf(0,0,0,0,0,0,0,0,0,0,2,2,2,2,2)
    	@JvmStatic val BINARY_CONTENT = gg.jte.runtime.BinaryContent.load(JtetagWithDefaultContentParamGenerated.javaClass, "JtetagWithDefaultContentParamGenerated.bin", 14,18);
    	@JvmStatic val TEXT_PART_BINARY_0 = BINARY_CONTENT.get(0)
    	@JvmStatic val TEXT_PART_BINARY_1 = BINARY_CONTENT.get(1)
    	@JvmStatic fun render(jteOutput:gg.jte.TemplateOutput, jteHtmlInterceptor:gg.jte.html.HtmlInterceptor?, one:Int = 1, content:gg.jte.Content) {
    		jteOutput.writeBinaryContent(TEXT_PART_BINARY_0)
    		jteOutput.writeUserContent(one)
    		jteOutput.writeBinaryContent(TEXT_PART_BINARY_1)
    		jteOutput.writeUserContent(content)
    	}
    	@JvmStatic fun renderMap(jteOutput:gg.jte.TemplateOutput, jteHtmlInterceptor:gg.jte.html.HtmlInterceptor?, params:Map<String, Any?>) {
    		val one = params["one"] as Int? ?: 1
    		val content = params["content"] as gg.jte.Content? ?: object : gg.jte.Content {
    			override fun writeTo(jteOutput:gg.jte.TemplateOutput) {
    				jteOutput.writeBinaryContent(TEXT_PART_BINARY_2)
    			}
    		}
    		render(jteOutput, jteHtmlInterceptor, one, content);
    	}
    }
}
```

The culprit is that line `jteOutput.writeBinaryContent(TEXT_PART_BINARY_2)` since there is only `TEXT_PART_BINARY_0` and `TEXT_PART_BINARY_1`. If I got it correctly, that code is generated here:

https://github.com/casid/jte/blob/e700eb01fe7272744e302075fc098b173cc66c77/jte-kotlin/src/main/java/gg/jte/compiler/kotlin/KotlinCodeGenerator.java#L254-L263

I'm not sure it should have been `binaryTextParts.size()` if `TEXT_PART_BINARY_` variables are 0-based. 🤔 

@casid, what do you think? Am I barking at the wrong tree?

Reply:
Hum, yeah, this seems bug. Also, I couldn't find a single test that mixes:

1. Binary rendering
2. Non-primitive parameters with non-null default values

These two tests:

https://github.com/casid/jte/blob/e700eb01fe7272744e302075fc098b173cc66c77/jte-kotlin/src/test/java/gg/jte/kotlin/TemplateEngine_HtmlOutputEscapingTest.java#L1003-L1019

They are almost there, but they are not running with `templateEngine.setBinaryStaticContent(true)`. I then created a subclass like this:

```java
package gg.jte.kotlin;

import org.junit.jupiter.api.BeforeEach;

public class TemplateEngine_HtmlOutputEscapingBinaryTest extends TemplateEngine_HtmlOutputEscapingTest {

    @BeforeEach
    void setUp() {
        templateEngine.setBinaryStaticContent(true);
    }
}
```

And... both tests fail. I don't have the bandwidth right now to fix the problem, but at least this investigation should point you to the right direction.

Reply:
I think you found a bug with default content parameters and binary templates. I'll try to have a look at it this weekend!

Reply:
@marcospereira thanks for finding this and for your investigation! I just pushed a fix for this to the main branch.

Reply:
Huh, this change fixed another bug, where line numbers where incorrectly displayed as 0 instead of the real one. Pushed another commit to adjust the tests for this.